In [1]:
#código de inicio
import pandas as pd
import numpy as np

# <img style="float: left; padding-right: 20px; width: 100px" src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/84/Escudo_de_la_Pontificia_Universidad_Cat%C3%B3lica_de_Chile.svg/1920px-Escudo_de_la_Pontificia_Universidad_Cat%C3%B3lica_de_Chile.svg.png"> MCD3020 - Introducción a Ciencia de Datos
**Pontificia Universidad Católica de Chile**<br>
**Magíster en Ciencia de Datos**<br>
**2022**<br>

----

# Tarea 2: Limpieza y combinación de datos.

***
## Instrucciones Generales:
- Esta Tarea debe ser desarrollada completamente en lenguaje de programación Python, en este mismo Notebook.
- El Notebook debe estar  ordenado, seguir buenas prácticas de escritura y programación, e incluir comentarios o celdas de markdown suficientes para explicar claramente todos lo códigos computacionales.
- El Notebook ya contiene algunas celdas marcadas con el comentario `#código de inicio`. Estas celdas han sido incluidas como ayuda para el desarrollo de la Tarea, y pueden ser ejecutadas tal como están.
- Las celdas marcadas como `#completar código` tienen un código parcial que debe ser completado para poder ser ejecutado. Ud debe agregar todas las líneas o bloques de código necesarios para desarrollar correctamente cada punto de la tarea. También puede eliminar estas celdas y partir el código desde cero si le resulta más conveniente.

***
## Introducción.

Hace ya casi 10 años, el trabajo de científico de datos fue catalogado por Harvard Bussiness Review como "el trabajo más atractivo del siglo XXI" [(Davenport & Patil 2012)](https://hbr.org/2012/10/data-scientist-the-sexiest-job-of-the-21st-century). Desde entonces, se ha comprobado un aumento constante de la demanda por profesionales expertos datos, y se espera que tanto la creación de puestos trabajos como los salarios sigan al alza en los próximos años. Los siguienes artículos de prensa y difusión ilustran esta situación:

https://www.smithhanley.com/2022/01/04/data-science-in-2022/
https://www.bbva.com/es/big-data-la-demanda-de-talento-experto-sigue-creciendo/

Los estudios citados hacen referencia a mercados laborales en Europa y Estados Unidos. Suponga que ud.está a cargo del desarrollo de un estudio del mercado laboral de científicos de datos en latinoamérica, para lo cual necesita construir, procesar y analizar una base de datos con las ofertas de trabajo publicadas en distintos países de la región.

En la **Tarea 1**, ud. avanzó en la primera etapa de construcción de la base de datos, utilizando web scraping para extraer los datos de ofertas de empleo en *data science* publicados en la red Linkedin, para una ciudad o país. Este proceso fue repetido y ampliado para todos los países de sudamérica, dando origen a un conjunto de bases de datos que serán su insumos para la siguiente etapa del proceso de ciencia de datos.  

Para considerar en el análisis el contexto económico y tecnológico de cada país, entre los datos se incluirán también otras variables como el *Índice global de innovación* (https://www.globalinnovationindex.org/Home), disponible online para distintos países del mundo.

En esta **Tarea 2**, ud. deberá integrar, limpiar y procesar las bases de datos obtenidas de distintas fuentes, de manera de obtener un conjunto de datos apto para distintos tipos de análisis.




## Datos de Entrada.

Los datos ofertas laborales de *data science* publicadas en Linkedin para los países de sudamérica están disponibles para lectura o descarga en los siguientes links:

* https://raw.githubusercontent.com/paguirre-uc/mds3020_2022/main/tarea2/jobsLinkedin_Argentina_t2.csv
* https://raw.githubusercontent.com/paguirre-uc/mds3020_2022/main/tarea2/jobsLinkedin_Bolivia_t2.csv
* https://raw.githubusercontent.com/paguirre-uc/mds3020_2022/main/tarea2/jobsLinkedin_Brasil_t2.csv
* https://raw.githubusercontent.com/paguirre-uc/mds3020_2022/main/tarea2/jobsLinkedin_Chile_t2.csv
* https://raw.githubusercontent.com/paguirre-uc/mds3020_2022/main/tarea2/jobsLinkedin_Colombia_t2.csv
* https://raw.githubusercontent.com/paguirre-uc/mds3020_2022/main/tarea2/jobsLinkedin_Guyana_t2.csv
* https://raw.githubusercontent.com/paguirre-uc/mds3020_2022/main/tarea2/jobsLinkedin_Ecuador_t2.csv
* https://raw.githubusercontent.com/paguirre-uc/mds3020_2022/main/tarea2/jobsLinkedin_Paraguay_t2.csv
* https://raw.githubusercontent.com/paguirre-uc/mds3020_2022/main/tarea2/jobsLinkedin_Peru_t2.csv
* https://raw.githubusercontent.com/paguirre-uc/mds3020_2022/main/tarea2/jobsLinkedin_Suriname_t2.csv
* https://raw.githubusercontent.com/paguirre-uc/mds3020_2022/main/tarea2/jobsLinkedin_Uruguay_t2.csv
* https://raw.githubusercontent.com/paguirre-uc/mds3020_2022/main/tarea2/jobsLinkedin_Venezuela_t2.csv

Además, se han descargado desde la página https://www.globalinnovationindex.org/analysis-indicator/ los datos de Índice de Innovación Global para el año 2022, disponibles en el siguiente archivo:

https://raw.githubusercontent.com/paguirre-uc/mds3020_2022/main/tarea2/indiceInnovacionPaises.csv

### 1. Lea todos los conjuntos de datos entregados, y concatenelos para generar un único dataframe con todas las ofertas laborales de Data Scientist publicadas en Linkedin para los países de sudamérica.

*Funciones útiles*: `pandas.concat()`

In [2]:
#completar codigo

locations = ['Chile','Argentina','Peru','Colombia','Ecuador','Brasil','Venezuela','Uruguay','Paraguay','Bolivia','Suriname','Guyana']

alldf=[]

for loc in locations:
    #url='https://raw.githubusercontent.com/paguirre-uc/mds3020_2022/main/tarea2/jobsLinkedin_%s_t2.csv'%loc
    ###
    ###
    alldf.append(##)
    
    
alldf=pd.concat(###)

SyntaxError: incomplete input (1761185492.py, line 14)

### Explicacion:


Como todos los csv tienen la misma ruta(excluyendo el pais), se cargan a alldf dentro de un for que solo recorre locations, asi completamos el nombre de cada csv y lo agregamos a alld. Luego con el concat concatenamos todo en un solo dataframe y se reinicia el indice

In [3]:
locations = ['Chile','Argentina','Peru','Colombia','Ecuador','Brasil','Venezuela','Uruguay','Paraguay','Bolivia','Suriname','Guyana']
alldf = []
for loc in locations:
    url = f'https://raw.githubusercontent.com/paguirre-uc/mds3020_2022/main/tarea2/jobsLinkedin_{loc}_t2.csv'
    df = pd.read_csv(url)
    alldf.append(df)
alldf = pd.concat(alldf, ignore_index=True)
alldf.head(5)


Unnamed: 0 Country                                       Location  \
0          0   Chile  Santiago, Santiago Metropolitan Region, Chile   
1          1   Chile  Santiago, Santiago Metropolitan Region, Chile   
2          2   Chile  Santiago, Santiago Metropolitan Region, Chile   
3          3   Chile              Rancagua, O'Higgins Region, Chile   
4          4   Chile  Santiago, Santiago Metropolitan Region, Chile   

             Title         Company  \
0   Data Scientist       BICE VIDA   
1   Data Scientist      BNamericas   
2  Data Scientist.  Fuerza Laboral   
3   Data Scientist       Agrosuper   
4   Data Scientist   MAS Analytics   

                                                 Url  \
0  https://cl.linkedin.com/jobs/view/data-scienti...   
1  https://cl.linkedin.com/jobs/view/data-scienti...   
2  https://cl.linkedin.com/jobs/view/data-scienti...   
3  https://cl.linkedin.com/jobs/view/data-scienti...   
4  https://cl.linkedin.com/jobs/view/data-scienti...   

                          Postulaciones  \
0                                   NaN   
1  \n          42 solicitudes\n           
2                                   NaN   
3                                   NaN   
4                                   NaN   

                                  Jornada  \
0  \n          Jornada completa\n           
1  \n          Jornada completa\n           
2  \n          Jornada completa\n           
3  \n          Jornada completa\n           
4  \n          Jornada completa\n           

                                  Antiguedad  
0   \n            No corresponde\n            
1  \n            Sin experiencia\n            
2  \n            Sin experiencia\n            
3       \n            Intermedio\n            
4   \n            No corresponde\n

### 2. ¿Existe alguna columna innecesaria o inútil en su dataframe? Justifique, y en caso necesario, elimine la(s) columna(s) de sobra.

In [130]:
alldf.columns

Index(['Unnamed: 0', 'Country', 'Location', 'Title', 'Company', 'Url',
       'Postulaciones', 'Jornada', 'Antiguedad'],
      dtype='object')

In [131]:
alldf['Unnamed: 0'].value_counts()

310    2
330    2
902    2
772    2
275    2
      ..
45     1
46     1
47     1
48     1
930    1
Name: Unnamed: 0, Length: 1000, dtype: int64

### Segun el dataframe, la columna q podriamos dropear es Unnamed:0 ya que no aporta información relevante. Además tampoco se especifica a que corresponde(Tampoco se puede inferir)

In [132]:
alldf = alldf.drop(columns=['Unnamed: 0'])
alldf

Country                                        Location  \
0     Argentina                            Greater Buenos Aires   
1     Argentina  Buenos Aires, Buenos Aires Province, Argentina   
2     Argentina  Buenos Aires, Buenos Aires Province, Argentina   
3     Argentina                Buenos Aires Province, Argentina   
4     Argentina  Buenos Aires, Buenos Aires Province, Argentina   
...         ...                                             ...   
1018  Venezuela                     Anzoategui State, Venezuela   
1019  Venezuela            Caracas, Federal District, Venezuela   
1020  Venezuela            Caracas, Federal District, Venezuela   
1021  Venezuela                     Anzoategui State, Venezuela   
1022  Venezuela                                       Venezuela   

                                                  Title  \
0                              Associate Data Scientist   
1                         Nerdearla 2022 - Data Science   
2                              Data Science - Data & AI   
3                                  ARG - Data Scientist   
4                               Data Scientist @Tiimiit   
...                                                 ...   
1018  Software Engineer, Trilogy (Remote) - $100,000...   
1019                    Senior Python Software Engineer   
1020                  Software Engineer - Data Platform   
1021  Software Engineer, Trilogy (Remote) - $100,000...   
1022                            software engineer mobil   

                           Company  \
0                BBVA en Argentina   
1              Accenture Argentina   
2              Accenture Argentina   
3     Movistar (Telefónica Hispam)   
4            Prisma Medios de Pago   
...                            ...   
1018            Crossover for Work   
1019                     Canonical   
1020                     Canonical   
1021            Crossover for Work   
1022                         TUTEN   

                                                    Url  \
0     https://ar.linkedin.com/jobs/view/associate-da...   
1     https://ar.linkedin.com/jobs/view/nerdearla-20...   
2     https://ar.linkedin.com/jobs/view/data-science...   
3     https://ar.linkedin.com/jobs/view/arg-data-sci...   
4     https://ar.linkedin.com/jobs/view/data-scienti...   
...                                                 ...   
1018  https://ve.linkedin.com/jobs/view/software-eng...   
1019  https://ve.linkedin.com/jobs/view/senior-pytho...   
1020  https://ve.linkedin.com/jobs/view/software-eng...   
1021  https://ve.linkedin.com/jobs/view/software-eng...   
1022  https://ve.linkedin.com/jobs/view/software-eng...   

                              Postulaciones  \
0                                       NaN   
1      \n          80 solicitudes\n           
2     \n          106 solicitudes\n           
3     \n          174 solicitudes\n           
4                                       NaN   
...                                     ...   
1018                                    NaN   
1019                                    NaN   
1020                                    NaN   
1021                                    NaN   
1022   \n          25 solicitudes\n           

                                     Jornada  \
0     \n          Jornada completa\n           
1     \n          Jornada completa\n           
2     \n          Jornada completa\n           
3     \n          Jornada completa\n           
4     \n          Jornada completa\n           
...                                      ...   
1018                                     NaN   
1019                                     NaN   
1020  \n          Jornada completa\n           
1021  \n          Jornada completa\n           
1022  \n          Jornada completa\n           

                                     Antiguedad  
0           \n            Ejecutivo\n            
1          \n            Intermedio\n            
2          \n            Intermedio\

### 3. Tratamiento de datos duplicados.
**a)** ¿Existen datos duplicados en su dataframe? En caso afirmativo, ¿cuántos son, y se trata de duplicados totales o parciales? Justifique.

**b)** En caso de existir duplicados, defina qué medida de corrección le parece apropiada (eliminar, corregir, combinar registros, etc), y aplíquela.

*Funciones útiles*: `pandas.duplicated()`,`pandas.drop_duplicates()`


In [133]:
duplicadosTotales = alldf.duplicated().sum()
print(f'Duplicados totales: {duplicadosTotales}')


Duplicados totales: 23


In [134]:
alldf = alldf.drop_duplicates()
print(f'Número de filas después de eliminar duplicados totales: {len(alldf)}')



Número de filas después de eliminar duplicados totales: 1000


#### Dataframe sin duplicados

In [135]:
alldf

Country                                        Location  \
0     Argentina                            Greater Buenos Aires   
1     Argentina  Buenos Aires, Buenos Aires Province, Argentina   
2     Argentina  Buenos Aires, Buenos Aires Province, Argentina   
3     Argentina                Buenos Aires Province, Argentina   
4     Argentina  Buenos Aires, Buenos Aires Province, Argentina   
...         ...                                             ...   
1016  Venezuela            Caracas, Federal District, Venezuela   
1017  Venezuela                     Anzoategui State, Venezuela   
1018  Venezuela                     Anzoategui State, Venezuela   
1019  Venezuela            Caracas, Federal District, Venezuela   
1020  Venezuela            Caracas, Federal District, Venezuela   

                                                  Title  \
0                              Associate Data Scientist   
1                         Nerdearla 2022 - Data Science   
2                              Data Science - Data & AI   
3                                  ARG - Data Scientist   
4                               Data Scientist @Tiimiit   
...                                                 ...   
1016        React/C# Fullstack Developer (LATAM Remote)   
1017  Software Developer, Trilogy (Remote) - $100,00...   
1018  Software Engineer, Trilogy (Remote) - $100,000...   
1019                    Senior Python Software Engineer   
1020                  Software Engineer - Data Platform   

                           Company  \
0                BBVA en Argentina   
1              Accenture Argentina   
2              Accenture Argentina   
3     Movistar (Telefónica Hispam)   
4            Prisma Medios de Pago   
...                            ...   
1016                     Qeystroke   
1017            Crossover for Work   
1018            Crossover for Work   
1019                     Canonical   
1020                     Canonical   

                                                    Url  \
0     https://ar.linkedin.com/jobs/view/associate-da...   
1     https://ar.linkedin.com/jobs/view/nerdearla-20...   
2     https://ar.linkedin.com/jobs/view/data-science...   
3     https://ar.linkedin.com/jobs/view/arg-data-sci...   
4     https://ar.linkedin.com/jobs/view/data-scienti...   
...                                                 ...   
1016  https://ve.linkedin.com/jobs/view/react-c%23-f...   
1017  https://ve.linkedin.com/jobs/view/software-dev...   
1018  https://ve.linkedin.com/jobs/view/software-eng...   
1019  https://ve.linkedin.com/jobs/view/senior-pytho...   
1020  https://ve.linkedin.com/jobs/view/software-eng...   

                              Postulaciones  \
0                                       NaN   
1      \n          80 solicitudes\n           
2     \n          106 solicitudes\n           
3     \n          174 solicitudes\n           
4                                       NaN   
...                                     ...   
1016                                    NaN   
1017                                    NaN   
1018                                    NaN   
1019                                    NaN   
1020                                    NaN   

                                     Jornada  \
0     \n          Jornada completa\n           
1     \n          Jornada completa\n           
2     \n          Jornada completa\n           
3     \n          Jornada completa\n           
4     \n          Jornada completa\n           
...                                      ...   
1016                                     NaN   
1017                                     NaN   
1018                                     NaN   
1019                                     NaN   
1020  \n          Jornada completa\n           

                                     Antiguedad  
0           \n            Ejecutivo\n            
1          \n            Intermedio\n            
2          \n            Intermedio\

In [137]:
duplicados_parciales = alldf.duplicated(subset=['Country', 'Location', 'Title', 'Company','Postulaciones','Jornada', 'Antiguedad']).sum()

print(f'Duplicados parciales: {duplicados_parciales}')


Duplicados parciales: 439


### Por lo tanto habian 23 filas duplicadas en el dataframe, las cuales decidi eliminar. Por otro lado, decidi no considerar la columna de url y obtuve un total de 439 filas repetidas parcialmente

### 4. Convierta la columna *Postulaciones* a un dato de tipo numérico.

**a)** Chequee los valores únicos de la columna *Postulaciones*.<br>
**b)** Realice las correcciones y operaciones de limpieza necesarias para convertir a tipo de dato numérico.<br>
**c)** Convierta la columna a datos de tipo `float`.<br>


*Funciones útiles*: `pandas.unique()`,`pandas.str.replace()`,`pandas.str.strip()`,`pandas.astype()`

In [138]:
#completar código
alldf['Postulaciones'].unique()

array([nan, '\n          80 solicitudes\n        ',
       '\n          106 solicitudes\n        ',
       '\n          174 solicitudes\n        ',
       '\n          53 solicitudes\n        ',
       '\n          32 solicitudes\n        ',
       '\n          85 solicitudes\n        ',
       '\n          43 solicitudes\n        ',
       '\n          44 solicitudes\n        ',
       '\n          47 solicitudes\n        ',
       '\n          33 solicitudes\n        ',
       '\n          55 solicitudes\n        ',
       '\n          58 solicitudes\n        ',
       '\n          65 solicitudes\n        ',
       '\n          79 solicitudes\n        ',
       '\n          34 solicitudes\n        ',
       '\n          59 applicants\n        ',
       '\n          94 applicants\n        ',
       '\n          60 applicants\n        ',
       '\n          191 applicants\n        ',
       '\n          129 applicants\n        ',
       '\n          151 applicants\n        ',
       '\

In [139]:

alldf['Postulaciones'] = alldf['Postulaciones'].str.replace(' solicitudes', '').str.replace(',', '').str.strip()
alldf['Postulaciones'] = pd.to_numeric(alldf['Postulaciones'], errors='coerce')
print(alldf.dtypes)


Country           object
Location          object
Title             object
Company           object
Url               object
Postulaciones    float64
Jornada           object
Antiguedad        object
dtype: object


C:\Users\ccana\AppData\Local\Temp\ipykernel_21192\318561036.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  alldf['Postulaciones'] = alldf['Postulaciones'].str.replace(' solicitudes', '').str.replace(',', '').str.strip()
C:\Users\ccana\AppData\Local\Temp\ipykernel_21192\318561036.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  alldf['Postulaciones'] = pd.to_numeric(alldf['Postulaciones'], errors='coerce')


In [142]:
alldf

Country                                        Location  \
0     Argentina                            Greater Buenos Aires   
1     Argentina  Buenos Aires, Buenos Aires Province, Argentina   
2     Argentina  Buenos Aires, Buenos Aires Province, Argentina   
3     Argentina                Buenos Aires Province, Argentina   
4     Argentina  Buenos Aires, Buenos Aires Province, Argentina   
...         ...                                             ...   
1016  Venezuela            Caracas, Federal District, Venezuela   
1017  Venezuela                     Anzoategui State, Venezuela   
1018  Venezuela                     Anzoategui State, Venezuela   
1019  Venezuela            Caracas, Federal District, Venezuela   
1020  Venezuela            Caracas, Federal District, Venezuela   

                                                  Title  \
0                              Associate Data Scientist   
1                         Nerdearla 2022 - Data Science   
2                              Data Science - Data & AI   
3                                  ARG - Data Scientist   
4                               Data Scientist @Tiimiit   
...                                                 ...   
1016        React/C# Fullstack Developer (LATAM Remote)   
1017  Software Developer, Trilogy (Remote) - $100,00...   
1018  Software Engineer, Trilogy (Remote) - $100,000...   
1019                    Senior Python Software Engineer   
1020                  Software Engineer - Data Platform   

                           Company  \
0                BBVA en Argentina   
1              Accenture Argentina   
2              Accenture Argentina   
3     Movistar (Telefónica Hispam)   
4            Prisma Medios de Pago   
...                            ...   
1016                     Qeystroke   
1017            Crossover for Work   
1018            Crossover for Work   
1019                     Canonical   
1020                     Canonical   

                                                    Url  Postulaciones  \
0     https://ar.linkedin.com/jobs/view/associate-da...            NaN   
1     https://ar.linkedin.com/jobs/view/nerdearla-20...           80.0   
2     https://ar.linkedin.com/jobs/view/data-science...          106.0   
3     https://ar.linkedin.com/jobs/view/arg-data-sci...          174.0   
4     https://ar.linkedin.com/jobs/view/data-scienti...            NaN   
...                                                 ...            ...   
1016  https://ve.linkedin.com/jobs/view/react-c%23-f...            NaN   
1017  https://ve.linkedin.com/jobs/view/software-dev...            NaN   
1018  https://ve.linkedin.com/jobs/view/software-eng...            NaN   
1019  https://ve.linkedin.com/jobs/view/senior-pytho...            NaN   
1020  https://ve.linkedin.com/jobs/view/software-eng...            NaN   

                                     Jornada  \
0     \n          Jornada completa\n           
1     \n          Jornada completa\n           
2     \n          Jornada completa\n           
3     \n          Jornada completa\n           
4     \n          Jornada completa\n           
...                                      ...   
1016                                     NaN   
1017                                     NaN   
1018                                     NaN   
1019                                     NaN   
1020  \n          Jornada completa\n           

                                     Antiguedad  
0           \n            Ejecutivo\n            
1          \n            Intermedio\n            
2          \n            Intermedio\n            
3          \n            Intermedio\n            
4      \n            No corresponde\n            
...                                         ...  
1016                                        NaN  
1017                                        NaN  
1018                                        NaN  
1019                                        NaN  
1020  \n            S

In [143]:
frec= alldf['Postulaciones'].value_counts()
frec

40.0     11
25.0      8
30.0      6
95.0      6
26.0      6
41.0      5
42.0      4
83.0      4
49.0      4
60.0      4
33.0      4
57.0      3
106.0     3
65.0      3
155.0     3
44.0      3
85.0      3
51.0      3
174.0     3
56.0      2
72.0      2
156.0     2
152.0     2
82.0      2
112.0     2
36.0      2
92.0      2
67.0      2
80.0      2
28.0      2
34.0      2
53.0      2
153.0     2
43.0      2
29.0      2
47.0      2
55.0      2
135.0     2
58.0      2
79.0      2
32.0      1
71.0      1
37.0      1
124.0     1
52.0      1
150.0     1
104.0     1
39.0      1
91.0      1
76.0      1
81.0      1
27.0      1
35.0      1
38.0      1
Name: Postulaciones, dtype: int64

In [144]:
frec= alldf['Postulaciones'].value_counts()
frec=frec.sum()
frec

142

In [145]:
cantidad_nan = alldf['Postulaciones'].isna().sum()
cantidad_nan

858

### 5. Convierta las columnas *Jornada* y *Antiguedad* a datos de tipo categórico.

**a)** Elimine los espacios en blanco sobrantes y caracteres de salto de línea ('\n') en ambas columnas.<br>
**b)** Chequee las categorías únicas de la columna *Jornada*. ¿Cuántas son? <br>
**c)** Para la columna *Antiguedad*, reduzca el número de categorías a sólo cuatro:`['N/A','Poca/ninguna','Intermedia','Senior/ejecutivo','Director']` (`N/A: no aplica`). <br>Para ello, defina un diccionario de equivalencias que le parezca adecuado.

**d)** Convierta ambas columnas a datos de tipo `category`.<br>

*Funciones útiles*: `pandas.unique()`,`pandas.str.strip()`,`pandas.map()`,`pandas.astype()`

In [146]:
alldf

Country                                        Location  \
0     Argentina                            Greater Buenos Aires   
1     Argentina  Buenos Aires, Buenos Aires Province, Argentina   
2     Argentina  Buenos Aires, Buenos Aires Province, Argentina   
3     Argentina                Buenos Aires Province, Argentina   
4     Argentina  Buenos Aires, Buenos Aires Province, Argentina   
...         ...                                             ...   
1016  Venezuela            Caracas, Federal District, Venezuela   
1017  Venezuela                     Anzoategui State, Venezuela   
1018  Venezuela                     Anzoategui State, Venezuela   
1019  Venezuela            Caracas, Federal District, Venezuela   
1020  Venezuela            Caracas, Federal District, Venezuela   

                                                  Title  \
0                              Associate Data Scientist   
1                         Nerdearla 2022 - Data Science   
2                              Data Science - Data & AI   
3                                  ARG - Data Scientist   
4                               Data Scientist @Tiimiit   
...                                                 ...   
1016        React/C# Fullstack Developer (LATAM Remote)   
1017  Software Developer, Trilogy (Remote) - $100,00...   
1018  Software Engineer, Trilogy (Remote) - $100,000...   
1019                    Senior Python Software Engineer   
1020                  Software Engineer - Data Platform   

                           Company  \
0                BBVA en Argentina   
1              Accenture Argentina   
2              Accenture Argentina   
3     Movistar (Telefónica Hispam)   
4            Prisma Medios de Pago   
...                            ...   
1016                     Qeystroke   
1017            Crossover for Work   
1018            Crossover for Work   
1019                     Canonical   
1020                     Canonical   

                                                    Url  Postulaciones  \
0     https://ar.linkedin.com/jobs/view/associate-da...            NaN   
1     https://ar.linkedin.com/jobs/view/nerdearla-20...           80.0   
2     https://ar.linkedin.com/jobs/view/data-science...          106.0   
3     https://ar.linkedin.com/jobs/view/arg-data-sci...          174.0   
4     https://ar.linkedin.com/jobs/view/data-scienti...            NaN   
...                                                 ...            ...   
1016  https://ve.linkedin.com/jobs/view/react-c%23-f...            NaN   
1017  https://ve.linkedin.com/jobs/view/software-dev...            NaN   
1018  https://ve.linkedin.com/jobs/view/software-eng...            NaN   
1019  https://ve.linkedin.com/jobs/view/senior-pytho...            NaN   
1020  https://ve.linkedin.com/jobs/view/software-eng...            NaN   

                                     Jornada  \
0     \n          Jornada completa\n           
1     \n          Jornada completa\n           
2     \n          Jornada completa\n           
3     \n          Jornada completa\n           
4     \n          Jornada completa\n           
...                                      ...   
1016                                     NaN   
1017                                     NaN   
1018                                     NaN   
1019                                     NaN   
1020  \n          Jornada completa\n           

                                     Antiguedad  
0           \n            Ejecutivo\n            
1          \n            Intermedio\n            
2          \n            Intermedio\n            
3          \n            Intermedio\n            
4      \n            No corresponde\n            
...                                         ...  
1016                                        NaN  
1017                                        NaN  
1018                                        NaN  
1019                                        NaN  
1020  \n            S

#### a

In [147]:

alldf['Jornada'] = alldf['Jornada'].str.strip()
alldf['Antiguedad'] = alldf['Antiguedad'].str.strip()


C:\Users\ccana\AppData\Local\Temp\ipykernel_21192\892637725.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  alldf['Jornada'] = alldf['Jornada'].str.strip()
C:\Users\ccana\AppData\Local\Temp\ipykernel_21192\892637725.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  alldf['Antiguedad'] = alldf['Antiguedad'].str.strip()


In [148]:
alldf.head(3)

Country                                        Location  \
0  Argentina                            Greater Buenos Aires   
1  Argentina  Buenos Aires, Buenos Aires Province, Argentina   
2  Argentina  Buenos Aires, Buenos Aires Province, Argentina   

                           Title              Company  \
0       Associate Data Scientist    BBVA en Argentina   
1  Nerdearla 2022 - Data Science  Accenture Argentina   
2       Data Science - Data & AI  Accenture Argentina   

                                                 Url  Postulaciones  \
0  https://ar.linkedin.com/jobs/view/associate-da...            NaN   
1  https://ar.linkedin.com/jobs/view/nerdearla-20...           80.0   
2  https://ar.linkedin.com/jobs/view/data-science...          106.0   

            Jornada  Antiguedad  
0  Jornada completa   Ejecutivo  
1  Jornada completa  Intermedio  
2  Jornada completa  Intermedio

#### b

In [151]:
categoriasJornada = alldf['Jornada'].unique()
cantidadCategorias = len(categoriasJornada)

print("Categorías únicas en la columna Jornada:", categoriasJornada)
print("Cantidad de categorías únicas:", cantidadCategorias)


Categorías únicas en la columna Jornada: ['Jornada completa' nan 'Contrato por obra' 'Full-time' 'Tempo integral'
 'Contract']
Cantidad de categorías únicas: 6


#### Aunque tecnicamente deberian de ser 5 las categorias unicas por que los tipo NaN no son realemente descriptivos, pero al ser muchos no los quise dropear

### c

In [152]:
categoriasAntiguedad = alldf['Antiguedad'].unique()
cantidadCategorias = len(categoriasAntiguedad)

print("Categorías únicas en la columna Antiguedad:", categoriasAntiguedad)
print("Cantidad de categorías únicas:", cantidadCategorias)

Categorías únicas en la columna Antiguedad: ['Ejecutivo' 'Intermedio' 'No corresponde' 'Sin experiencia' nan
 'Algo de responsabilidad' 'Jornada completa' 'Associate' 'Entry level'
 'Not Applicable' 'Mid-Senior level' 'Director' 'Assistente'
 'Não aplicável' 'Pleno-sênior' 'Júnior' 'Full-time' 'Media jornada']
Cantidad de categorías únicas: 18


In [154]:
categoriasAntiguedad.tolist()

['Ejecutivo',
 'Intermedio',
 'No corresponde',
 'Sin experiencia',
 nan,
 'Algo de responsabilidad',
 'Jornada completa',
 'Associate',
 'Entry level',
 'Not Applicable',
 'Mid-Senior level',
 'Director',
 'Assistente',
 'Não aplicável',
 'Pleno-sênior',
 'Júnior',
 'Full-time',
 'Media jornada']

In [199]:

mapeoAntiguedad = {
    'N/A': ['No corresponde', 'Not Applicable', 'Não aplicável', 'Media jornada', 'Jornada completa', 'Full-time'],
    'Poca/ninguna': ['Sin experiencia', 'Algo de responsabilidad', 'Entry level', 'Júnior', 'Assistente'],
    'Intermedia': ['Intermedio', 'Associate'],
    'Senior/ejecutivo': ['Ejecutivo', 'Mid-Senior level', 'Pleno-sênior'],
    'Director': ['Director']
}

for nuevaCategoria, categoriasActuales in mapeoAntiguedad.items():
    alldf.loc[alldf['Antiguedad'].isin(categoriasActuales), 'Antiguedad'] = nuevaCategoria

categoriasAntiguedad = alldf['Antiguedad'].unique()
print("Categorías únicas en la columna Antiguedad después del cambio:")
print(categoriasAntiguedad)


Categorías únicas en la columna Antiguedad después del cambio:
['Senior/ejecutivo', 'Intermedia', 'N/A', 'Poca/ninguna', NaN, 'Director']
Categories (5, object): ['Director', 'Intermedia', 'N/A', 'Poca/ninguna', 'Senior/ejecutivo']


In [156]:
alldf

Country                                        Location  \
0     Argentina                            Greater Buenos Aires   
1     Argentina  Buenos Aires, Buenos Aires Province, Argentina   
2     Argentina  Buenos Aires, Buenos Aires Province, Argentina   
3     Argentina                Buenos Aires Province, Argentina   
4     Argentina  Buenos Aires, Buenos Aires Province, Argentina   
...         ...                                             ...   
1016  Venezuela            Caracas, Federal District, Venezuela   
1017  Venezuela                     Anzoategui State, Venezuela   
1018  Venezuela                     Anzoategui State, Venezuela   
1019  Venezuela            Caracas, Federal District, Venezuela   
1020  Venezuela            Caracas, Federal District, Venezuela   

                                                  Title  \
0                              Associate Data Scientist   
1                         Nerdearla 2022 - Data Science   
2                              Data Science - Data & AI   
3                                  ARG - Data Scientist   
4                               Data Scientist @Tiimiit   
...                                                 ...   
1016        React/C# Fullstack Developer (LATAM Remote)   
1017  Software Developer, Trilogy (Remote) - $100,00...   
1018  Software Engineer, Trilogy (Remote) - $100,000...   
1019                    Senior Python Software Engineer   
1020                  Software Engineer - Data Platform   

                           Company  \
0                BBVA en Argentina   
1              Accenture Argentina   
2              Accenture Argentina   
3     Movistar (Telefónica Hispam)   
4            Prisma Medios de Pago   
...                            ...   
1016                     Qeystroke   
1017            Crossover for Work   
1018            Crossover for Work   
1019                     Canonical   
1020                     Canonical   

                                                    Url  Postulaciones  \
0     https://ar.linkedin.com/jobs/view/associate-da...            NaN   
1     https://ar.linkedin.com/jobs/view/nerdearla-20...           80.0   
2     https://ar.linkedin.com/jobs/view/data-science...          106.0   
3     https://ar.linkedin.com/jobs/view/arg-data-sci...          174.0   
4     https://ar.linkedin.com/jobs/view/data-scienti...            NaN   
...                                                 ...            ...   
1016  https://ve.linkedin.com/jobs/view/react-c%23-f...            NaN   
1017  https://ve.linkedin.com/jobs/view/software-dev...            NaN   
1018  https://ve.linkedin.com/jobs/view/software-eng...            NaN   
1019  https://ve.linkedin.com/jobs/view/senior-pytho...            NaN   
1020  https://ve.linkedin.com/jobs/view/software-eng...            NaN   

               Jornada        Antiguedad  
0     Jornada completa  Senior/ejecutivo  
1     Jornada completa        Intermedia  
2     Jornada completa        Intermedia  
3     Jornada completa        Intermedia  
4     Jornada completa               N/A  
...                ...               ...  
1016               NaN               NaN  
1017               NaN               NaN  
1018               NaN               NaN  
1019               NaN               NaN  
1020  Jornada completa      Poca/ninguna  

[1000 rows x 8 columns]

In [157]:
categoriasAntiguedad = alldf['Antiguedad'].unique()
cantidadCategorias = len(categoriasAntiguedad)

print("Categorías únicas en la columna Antiguedad:", categoriasAntiguedad)
print("Cantidad de categorías únicas:", cantidadCategorias)

Categorías únicas en la columna Antiguedad: ['Senior/ejecutivo' 'Intermedia' 'N/A' 'Poca/ninguna' nan 'Director']
Cantidad de categorías únicas: 6


### d

In [158]:
alldf['Jornada'] = alldf['Jornada'].astype('category')
alldf['Antiguedad'] = alldf['Antiguedad'].astype('category')

print("Tipos de datos después de la conversión:")
print(alldf.dtypes)

Tipos de datos después de la conversión:
Country            object
Location           object
Title              object
Company            object
Url                object
Postulaciones     float64
Jornada          category
Antiguedad       category
dtype: object


C:\Users\ccana\AppData\Local\Temp\ipykernel_21192\1498524546.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  alldf['Jornada'] = alldf['Jornada'].astype('category')
C:\Users\ccana\AppData\Local\Temp\ipykernel_21192\1498524546.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  alldf['Antiguedad'] = alldf['Antiguedad'].astype('category')


In [159]:
alldf.head(5)

Country                                        Location  \
0  Argentina                            Greater Buenos Aires   
1  Argentina  Buenos Aires, Buenos Aires Province, Argentina   
2  Argentina  Buenos Aires, Buenos Aires Province, Argentina   
3  Argentina                Buenos Aires Province, Argentina   
4  Argentina  Buenos Aires, Buenos Aires Province, Argentina   

                           Title                       Company  \
0       Associate Data Scientist             BBVA en Argentina   
1  Nerdearla 2022 - Data Science           Accenture Argentina   
2       Data Science - Data & AI           Accenture Argentina   
3           ARG - Data Scientist  Movistar (Telefónica Hispam)   
4        Data Scientist @Tiimiit         Prisma Medios de Pago   

                                                 Url  Postulaciones  \
0  https://ar.linkedin.com/jobs/view/associate-da...            NaN   
1  https://ar.linkedin.com/jobs/view/nerdearla-20...           80.0   
2  https://ar.linkedin.com/jobs/view/data-science...          106.0   
3  https://ar.linkedin.com/jobs/view/arg-data-sci...          174.0   
4  https://ar.linkedin.com/jobs/view/data-scienti...            NaN   

            Jornada        Antiguedad  
0  Jornada completa  Senior/ejecutivo  
1  Jornada completa        Intermedia  
2  Jornada completa        Intermedia  
3  Jornada completa        Intermedia  
4  Jornada completa               N/A

### 6. Lea el conjunto de datos de Índice de Innovación por país y renombre la columna *Economy* a *Country*.


In [160]:
url = "https://raw.githubusercontent.com/paguirre-uc/mds3020_2022/main/tarea2/indiceInnovacionPaises.csv"
df = pd.read_csv(url)
df.head(5)


rank                   Economy Income Group(Strength/Weakness)  \
0     1               Switzerland                        Strength   
1     2  United States of America                        Strength   
2     3                    Sweden                        Strength   
3     4            United Kingdom                        Strength   
4     5               Netherlands                             NaN   

  Strength / Weakness  Score  
0            Strength   64.6  
1            Strength   61.8  
2            Strength   61.6  
3            Strength   59.7  
4            Strength   58.0

In [161]:
df.rename(columns={'Economy': 'Country'}, inplace=True)

In [164]:
df

rank                   Country Income Group(Strength/Weakness)  \
0       1               Switzerland                        Strength   
1       2  United States of America                        Strength   
2       3                    Sweden                        Strength   
3       4            United Kingdom                        Strength   
4       5               Netherlands                             NaN   
..    ...                       ...                             ...   
127   128                     Yemen                             NaN   
128   129                Mauritania                        Weakness   
129   130                   Burundi                        Weakness   
130   131                      Iraq                        Weakness   
131   132                    Guinea                        Weakness   

    Strength / Weakness  Score  
0              Strength   64.6  
1              Strength   61.8  
2              Strength   61.6  
3              Strength   59.7  
4              Strength   58.0  
..                  ...    ...  
127                 NaN   13.8  
128                 NaN   12.4  
129                 NaN   12.3  
130            Weakness   11.9  
131            Weakness   11.6  

[132 rows x 5 columns]

### 7. Combinación de conjuntos de datos.
**a)** Del dataframe de Índice de Innovación por País, seleccione sólo las columnas *Country* y *Score*. <br>
**b)** Combine estos datos con el dataframe de ofertas laborales, de manera de agregarle a éste una columna *Score* con el índice de innovación del país dónde se publicó el anuncio: <br>

|Country | Location|Title | Company | Url | Postulaciones | Jornada |Antiguedad |Score |
|---|---|---|---|---|---|---|---|---|

Este dataframe debe tener el mismo número de registros que el dataframe de ofertas laborales.

**b)** Determine si en este dataframe final, existen registros para los cuales la columna *Score* es un dato faltante o nulo. ¿A qué se deben estos datos nulos, si es que existen?

*Funciones útiles*: `pandas.merge()`

### a

In [173]:
dfFiltrado = df[['Country', 'Score']].copy()
dfFiltrado


Country  Score
0                 Switzerland   64.6
1    United States of America   61.8
2                      Sweden   61.6
3              United Kingdom   59.7
4                 Netherlands   58.0
..                        ...    ...
127                     Yemen   13.8
128                Mauritania   12.4
129                   Burundi   12.3
130                      Iraq   11.9
131                    Guinea   11.6

[132 rows x 2 columns]

In [174]:
dfFinal = pd.merge(alldf, dfFiltrado, on='Country', how='left')
dfFinal


Country                                        Location  \
0    Argentina                            Greater Buenos Aires   
1    Argentina  Buenos Aires, Buenos Aires Province, Argentina   
2    Argentina  Buenos Aires, Buenos Aires Province, Argentina   
3    Argentina                Buenos Aires Province, Argentina   
4    Argentina  Buenos Aires, Buenos Aires Province, Argentina   
..         ...                                             ...   
995  Venezuela            Caracas, Federal District, Venezuela   
996  Venezuela                     Anzoategui State, Venezuela   
997  Venezuela                     Anzoategui State, Venezuela   
998  Venezuela            Caracas, Federal District, Venezuela   
999  Venezuela            Caracas, Federal District, Venezuela   

                                                 Title  \
0                             Associate Data Scientist   
1                        Nerdearla 2022 - Data Science   
2                             Data Science - Data & AI   
3                                 ARG - Data Scientist   
4                              Data Scientist @Tiimiit   
..                                                 ...   
995        React/C# Fullstack Developer (LATAM Remote)   
996  Software Developer, Trilogy (Remote) - $100,00...   
997  Software Engineer, Trilogy (Remote) - $100,000...   
998                    Senior Python Software Engineer   
999                  Software Engineer - Data Platform   

                          Company  \
0               BBVA en Argentina   
1             Accenture Argentina   
2             Accenture Argentina   
3    Movistar (Telefónica Hispam)   
4           Prisma Medios de Pago   
..                            ...   
995                     Qeystroke   
996            Crossover for Work   
997            Crossover for Work   
998                     Canonical   
999                     Canonical   

                                                   Url  Postulaciones  \
0    https://ar.linkedin.com/jobs/view/associate-da...            NaN   
1    https://ar.linkedin.com/jobs/view/nerdearla-20...           80.0   
2    https://ar.linkedin.com/jobs/view/data-science...          106.0   
3    https://ar.linkedin.com/jobs/view/arg-data-sci...          174.0   
4    https://ar.linkedin.com/jobs/view/data-scienti...            NaN   
..                                                 ...            ...   
995  https://ve.linkedin.com/jobs/view/react-c%23-f...            NaN   
996  https://ve.linkedin.com/jobs/view/software-dev...            NaN   
997  https://ve.linkedin.com/jobs/view/software-eng...            NaN   
998  https://ve.linkedin.com/jobs/view/senior-pytho...            NaN   
999  https://ve.linkedin.com/jobs/view/software-eng...            NaN   

              Jornada        Antiguedad  Score  
0    Jornada completa  Senior/ejecutivo   28.6  
1    Jornada completa        Intermedia   28.6  
2    Jornada completa        Intermedia   28.6  
3    Jornada completa        Intermedia   28.6  
4    Jornada completa               N/A   28.6  
..                ...               ...    ...  
995               NaN               NaN    NaN  
996               NaN               NaN    NaN  
997               NaN               NaN    NaN  
998               NaN               NaN    NaN  
999  Jornada completa      Poca/ninguna    NaN  

[1000 rows x 9 columns]

### b

In [175]:
nulos = dfFinal[dfFinal['Score'].isnull()]
if nulos.empty:
    print("No hay datos con Score nulo.")
else:
    print(f"Hay {len(nulos)} datos con Score nulo.")
 


Hay 485 registros con Score nulo.


In [187]:
paises= dfFiltrado['Country'].unique().tolist()
paisesOrdenadosdfFinal = sorted(paises)
paisesOrdenadosdfFinal


['Albania',
 'Algeria',
 'Angola',
 'Argentina',
 'Armenia',
 'Australia',
 'Austria',
 'Azerbaijan',
 'Bahrain',
 'Bangladesh',
 'Belarus',
 'Belgium',
 'Benin',
 'Bosnia and Herzegovina',
 'Botswana',
 'Brazil',
 'Brunei Darussalam',
 'Bulgaria',
 'Burkina Faso',
 'Burundi',
 "C?te d'Ivoire",
 'Cambodia',
 'Cameroon',
 'Canada',
 'Chile',
 'China',
 'Colombia',
 'Costa Rica',
 'Croatia',
 'Cyprus',
 'Czech Republic',
 'Denmark',
 'Dominican Republic',
 'Ecuador',
 'Egypt',
 'El Salvador',
 'Estonia',
 'Ethiopia',
 'Finland',
 'France',
 'Georgia',
 'Germany',
 'Ghana',
 'Greece',
 'Guatemala',
 'Guinea',
 'Honduras',
 'Hong Kong',
 'Hungary',
 'Iceland',
 'India',
 'Indonesia',
 'Iran (Islamic Republic of)',
 'Iraq',
 'Ireland',
 'Israel',
 'Italy',
 'Jamaica',
 'Japan',
 'Jordan',
 'Kazakhstan',
 'Kenya',
 'Kuwait',
 'Kyrgyzstan',
 "Lao People's Democratic Republic",
 'Latvia',
 'Lithuania',
 'Luxembourg',
 'Madagascar',
 'Malaysia',
 'Mali',
 'Malta',
 'Mauritania',
 'Mauritius',
 

In [189]:
paises= alldf['Country'].unique().tolist()
paisesOrdenadosAlldf = sorted(paises)
paisesOrdenadosAlldf


['Argentina',
 'Bolivia',
 'Brasil',
 'Chile',
 'Colombia',
 'Ecuador',
 'Guyana',
 'Paraguay',
 'Suriname',
 'Uruguay',
 'Venezuela']

In [194]:
paisesOrdenadosdfFinal = dfFiltrado['Country'].unique().tolist()
paisesOrdenadosAlldf = alldf['Country'].unique().tolist()
paises_dfFinal = set(paisesOrdenadosdfFinal)
paises_alldf = set(paisesOrdenadosAlldf)
paisesQueNoEstan = paises_alldf - paises_dfFinal

print("Países de alldf que no están en dfFiltrado:")
print(paisesQueNoEstan)



Países de alldf que no están en dfFiltrado:
{'Suriname', 'Brasil', 'Venezuela', 'Bolivia', 'Guyana'}


### Según lo señalado anteriormente una de la razones del por que hay valores NaN es por que algunos paises no se encontraban en el dataframe que entrega el score. Como consideración, se podria limpiar aún más los datos del dataframe. Como signos del tipo - , . en nombres 